pkg/capacity/manager.go

```diff
@@ -6,57 +6,38 @@ import (
 	"sort"
 
 	toolchainv1alpha1 "github.com/codeready-toolchain/api/api/v1alpha1"
-	"github.com/codeready-toolchain/host-operator/controllers/toolchainconfig"
 	"github.com/codeready-toolchain/host-operator/pkg/counter"
 	"github.com/codeready-toolchain/toolchain-common/pkg/cluster"
 	"github.com/codeready-toolchain/toolchain-common/pkg/condition"
 
-	"k8s.io/apimachinery/pkg/types"
 	runtimeclient "sigs.k8s.io/controller-runtime/pkg/client"
 )
 
 type (
 	spaceProvisionerConfigPredicate func(*toolchainv1alpha1.SpaceProvisionerConfig) bool
-)
 
-func hasNotReachedMaxNumberOfSpacesThreshold(counts counter.Counts) spaceProvisionerConfigPredicate {
-	return func(spc *toolchainv1alpha1.SpaceProvisionerConfig) bool {
-		numberOfSpaces := uint(counts.SpacesPerClusterCounts[spc.Spec.ToolchainCluster])
-		threshold := spc.Spec.CapacityThresholds.MaxNumberOfSpaces
-		return threshold == 0 || numberOfSpaces < threshold
-	}
-}
+	// SpaceCountGetter is a function useful for mocking the counts cache and can be passed
+	// to the NewClusterManager function. The returned tuple represents the actual number
+	// of the spaces found for given cluster and the fact whether the value was actually found or not
+	// (similar to the return value of the map lookup).
+	SpaceCountGetter func(ctx context.Context, clusterName string) (int, bool)
+)
 
-func hasEnoughMemoryCapacity(status *toolchainv1alpha1.ToolchainStatus) spaceProvisionerConfigPredicate {
+// Even though the readiness status of the SpaceProvisionerConfig is based (in part) on the space count
+// we do another check here using the fresh data from the counts cache. This is to ensure we don't
+// overcommit spaces to the clusters.
+// The returned predicate assumes that the incoming SPC has already been updated with the fresh value
+// from the counts cache.
+func checkHasNotReachedSpaceCountThreshold() spaceProvisionerConfigPredicate {
 	return func(spc *toolchainv1alpha1.SpaceProvisionerConfig) bool {
-		threshold := spc.Spec.CapacityThresholds.MaxMemoryUtilizationPercent
-		if threshold == 0 {
-			return true
-		}
-		for _, memberStatus := range status.Status.Members {
-			if memberStatus.ClusterName == spc.Spec.ToolchainCluster {
-				return hasMemberEnoughMemoryCapacity(memberStatus, threshold)
-			}
+		var spaceCount int
+		if spc.Status.ConsumedCapacity != nil {
+			spaceCount = spc.Status.ConsumedCapacity.SpaceCount
```

Comment:
correct me if I'm wrong, but we shouldn't use the SPC for getting the number of provisioned Spaces

Reply:
This has been reworked. A custom struct is now used in the algorithm so the data flow is more clear.